Add protobuf based Encoder and Decoder

cw9 opened the pull request:
- Added protobuf based Encoder and Decoder.
- Added a connection based EncodeDecoder which will be used by consumers.

@xichen2020 @schallert @jeromefroe

Commit:
Add protobuf based protocol

Commit:
Add comments

xichen2020 commented on generated/proto/msgpb/msg.proto:
```diff
@@ -0,0 +1,16 @@
+package msgpb;
+
+message Header {
+    uint64 high = 1;
```
More descriptive names?

cw9 replied:
sounds good

xichen2020 commented on protocol/proto/addr_enc_dec.go:
```diff
@@ -0,0 +1,254 @@
+		// Should retry forever on reconnect.
+		retrier:             retry.NewRetrier(opts.ConnectionRetryOptions().SetForever(defaultRetryForever)),
+		dialTimeout:         opts.DialTimeout(),
+		reconnectDelayNanos: opts.ReconnectDelay().Nanoseconds(),
```
Hm should this be part of the retry options?

cw9 replied:
This is a different delay, it avoids signaling connection reset while a reset is ongoing, it's a nice-to-have, not required.

xichen2020 commented on protocol/proto/conn_enc_dec.go:
```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
```
Feels like this `ConnectionEncoderDecoder` interface isn't all that useful. Seems like all it does atm is just delegating the encoding/decoding logic to the actual encoder/decoder. Might be cleaner to merge what's in here to the `AddressEncoderDecoder` to reduce cognitive burden.

cw9 replied:
It allows resetting connection so it can be reused and pooled which is ideal for the aggregator consumer use case, it does not need the logic to retry tcp dials, which is much simpler.  
AddressEncodeDecoder have the extra logic to retry connections which is only needed for producer cases. Originally I had them together but felt a separation is much cleaner.

xichen2020 commented on protocol/proto/addr_enc_dec.go:
```diff
@@ -0,0 +1,254 @@
+	resetConnError tally.Counter
+}
+
+// NewAddressEncodeDecoder creates an EncodeDecoder that
```
Hm right now it feels the encoding/decoding logic feels very entangled with connection management logic, any reason not to have an underlying `RetryableTCPConnection` interface that takes care of the retry logic etc (e.g., https://github.com/m3db/m3collector/blob/master/backend/msgpack/conn.go) so the higher-level encoder / decoder implementation only needs to focus on encoding/decoding data?

cw9 replied:
This AddressEncodeDecode is basically a combination of retryable connection + EncodeDecoder? And it just calls the EncodeDecoder for real encode/decode work?

xichen2020 commented on protocol/proto/addr_enc_dec.go:
```diff
@@ -0,0 +1,254 @@
+}
+
+func (c *addrEncdec) Encode(msg Marshaler) error {
+	if !c.hasValidConnection() {
```
Do you need to take a read lock during encoding? Otherwise there's a race condition where you might still encode after the object is closed.

cw9 replied:
Possibly, can put this in a read lock

xichen2020 commented on protocol/proto/addr_enc_dec.go:
```diff
@@ -0,0 +1,254 @@
+	return nil
+}
+
+func (c *addrEncdec) Decode(acks Unmarshaler) error {
```
Do you need to take a read lock during decoding? Otherwise there's a race condition where you might still decode after the object is closed.

cw9 replied:
Possibly, can put this in a read lock

xichen2020 commented on protocol/proto/addr_enc_dec.go:
```diff
@@ -0,0 +1,254 @@
+	c.validConn.Store(false)
+	c.encdec.Close()
+	close(c.doneCh)
+	c.Unlock()
```
Would suggest `defer` here since this is likely not going to be on the hot path.

cw9 replied:
sounds good

xichen2020 commented on protocol/proto/decode.go:
```diff
@@ -0,0 +1,81 @@
+	if _, err := io.ReadFull(d.r, d.sizeBuffer); err != nil {
+		return 0, err
+	}
+	size := binary.BigEndian.Uint32(d.sizeBuffer)
```
Use a shared var for endianness during encoding/decoding?

cw9 replied:
sounds good

xichen2020 commented on protocol/proto/encode.go:
```diff
@@ -0,0 +1,100 @@
+	e.w.Reset(w)
+}
+
+func growDataBufferIfNeeded(buffer []byte, target int, pool pool.BytesPool) []byte {
```
Pull this into a common file `common.go`? This is also used in `decode.go`.

cw9 replied:
sounds good

Commit:
Address comments

Commit:
Update proto

Commit:
Update proto name

Commit:
Use defer in test routines

Commit:
Remove address based encode decoder

cw9 commented:
Removed the old AddressEncodeDecoder out of protocol package as the connection management logic is not really about encode/decode protocol and should live with producer package.

Renamed the ConnectionEncodeDecoder to just EncodeDecoder

@xichen2020

Commit:
Remove lock

Commit:
Handle nil connections safely

xichen2020 commented on protocol/proto/base_decoder.go:
```diff
@@ -0,0 +1,80 @@
+	return newDecoder(r, opts)
+}
+
+func newDecoder(r io.Reader, opts BaseOptions) *decoder {
```
Why both `NewDecoder` and `newDecoder`?

cw9 replied:
Cause I did not expose resetReader on the Decoder interface and is only available on *decoder and will be used by the EncodeDecoder, I want user to only reset both writer and reader through EncodeDecoder rather than allow them to reset separately, but I'm open to it.

xichen2020 commented on protocol/proto/base_decoder.go:
```diff
@@ -0,0 +1,80 @@
+	}
+	pool := opts.BytesPool()
+	return &decoder{
+		r:          bufio.NewReaderSize(r, opts.BufferSize()),
```
Do you want to special case if `r` is already a `*bufio.Reader`?

cw9 replied:
I think NewReaderSize already handles that

xichen2020 commented on protocol/proto/base_decoder.go:
```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
```
Just call this `decoder.go` unless you have a more concrete decoder implementation in mind?

cw9 replied:
sure

xichen2020 commented on protocol/proto/base_encode.go:
```diff
@@ -0,0 +1,80 @@
+}
+
+// NewEncoder creates a new encoder.
+func NewEncoder(w io.Writer, opts BaseOptions) Encoder {
```
Why both `NewEncoder` and `newEncoder`?

cw9 replied:
same, to not expose resetWriter()

xichen2020 commented on protocol/proto/base_encode.go:
```diff
@@ -0,0 +1,80 @@
+}
+
+func (e *encoder) encodeSize(size int) error {
+	sizeEncodeDecoder.PutUint32(e.sizeBuffer, uint32(size))
```
Do you want to encode size as varint? I'd imagine most sizes are fairly small.

cw9 replied:
We could potentially do some further benchmarking on comparing varint vs int32, but this is the way grpc streams protobuf as well, I think it should work reasonably well. Could keep an eye on varint in the future.

xichen2020 commented on protocol/proto/common.go:
```diff
@@ -0,0 +1,52 @@
+
+var sizeEncodeDecoder = binary.BigEndian
+
+func growDataBufferIfNeeded(buffer []byte, target int, pool pool.BytesPool) []byte {
```
nit: `target` -> `targetSize`

cw9 replied:
sure

xichen2020 commented on protocol/proto/common.go:
```diff
@@ -0,0 +1,52 @@
+	}
+	b := pool.Get(target)
+	// Make sure there is enough length in the slice.
+	return b[:cap(b)]
```
Hm should this be `b[:target]`?

cw9 replied:
it's better to do cap, for example the pool has {10, 512, 1024} buckets, and we did a Get(100), it will return a slice with 512 cap, but next time when we want to grow the slice to 200, we need to check the cap and reset the len, I feel it's simpler to just get the full cap in the first time.

xichen2020 replied:
Hm in that case should you compare `cap(buffer)` against `targetSize` then?

cw9 replied:
in this case len == cap, it's safer to compare len with targetSize no? in case the len is not at cap and the user may want to access idx that is between [len, cap)

xichen2020 commented on protocol/proto/encode_decoder.go:
```diff
@@ -0,0 +1,91 @@
+)
+
+var (
+	defaultConn = new(net.TCPConn)
```
Hm why is this necessary?

cw9 replied:
In case user initialize the EncodeDecoder with a nil conn, in a consumer writer where it needs to talk to a target address, the target might not be available (host being down) so it will need a background thread to connect to the address while the conn here is nil, I guess we could do stuff to guard writing when the conn is nil, but safer to guard here as well to avoid panics.

cw9 replied:
Removed this for now, could just defend this from user side

xichen2020 commented on protocol/proto/encode_decoder.go:
```diff
@@ -0,0 +1,91 @@
+	defaultConn = new(net.TCPConn)
+)
+
+type encdec struct {
```
Is this meant to be thread-unsafe?

cw9 replied:
Yes, all the encode decode are thread unsafe.

xichen2020 replied:
Perhaps add a comment above the struct declaration?

xichen2020 commented on protocol/proto/encode_decoder.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+func (c *encdec) Encode(msg Marshaler) error {
+	err := c.enc.Encode(msg)
```
nit: return ...

cw9 replied:
sure

xichen2020 commented on protocol/proto/encode_decoder.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+func (c *encdec) Decode(acks Unmarshaler) error {
+	err := c.dec.Decode(acks)
```
nit: return ...

cw9 replied:
sure

xichen2020 commented on protocol/proto/encode_decoder.go:
```diff
@@ -0,0 +1,91 @@
+
+func (c *encdec) Reset(conn net.Conn) {
+	if conn == nil {
+		conn = defaultConn
```
I'm not sure why this is necessary? We should never call `Reset` and ctor with a nil conn.

cw9 replied:
Just in case of bad behaviors

xichen2020 commented on protocol/proto/encode_decoder_test.go:
```diff
@@ -0,0 +1,71 @@
+		Value: make([]byte, 10),
+	}
+	go func() {
+		require.NoError(t, c.Encode(&testMsg))
```
Do you need to wait for this to finish before returning from the test?

cw9 replied:
The decode is essentially waiting for this to finish

xichen2020 commented on protocol/proto/options.go:
```diff
@@ -0,0 +1,105 @@
+)
+
+const (
+	defaultDataBufferSize = 16384
```
Why 16K?

cw9 replied:
From a benchmark I saw 16k behaves much faster than just 1 few ks, and minimal difference between 16k to 64k, so I decided to go with 16k. gRPC by default is using 32K.

xichen2020 replied:
Sounds good, can you add a comment including the rationale? Might be good to include the benchmarks if possible.

xichen2020 commented on protocol/proto/types.go:
```diff
@@ -0,0 +1,111 @@
+// Marshaler can be marshaled.
+type Marshaler interface {
+	Size() int
+	MarshalTo(data []byte) (int, error)
```
Should provide functional comments for the APIs? Doesn't look like these methods are documented either here or in the implementation.

cw9 replied:
Sure can add comments

xichen2020 commented on protocol/proto/types.go:
```diff
@@ -0,0 +1,111 @@
+	Close()
+
+	// Reset resets the EncodeDecoder.
+	Reset(conn net.Conn)
```
Do you need a connection? Can we use `io.ReadWriter`? Seems a bit weird to have a connection in the encoder decoder interface.

cw9 replied:
Yeah can do that

Commit:
Address comments

Commit:
Remove default tcp conn

Commit:
Add comment

xichen2020 reviewed (approved):
LGTM
